File: sdk/typescript/src/api/webhooks.ts
/**
 * Webhooks API
 *
 * Modular webhooks API for use with HttpClient.
 */

import { HttpClient } from '../http';
import type { WebhookEndpoint, WebhookEventType, ListResponse } from '../types';

/**
 * Webhooks API for managing webhook endpoints.
 */
export class WebhooksAPI {
  constructor(private readonly http: HttpClient) {}

  /**
   * Create a webhook endpoint.
   *
   * @param params - Webhook configuration
   * @returns Created webhook endpoint with secret
   *
   * @example
   * ```typescript
   * const webhook = await webhooks.create({
   *   url: 'https://myserver.com/webhooks/execution-market',
   *   events: ['task.completed', 'task.submitted']
   * });
   * // Save webhook.secret for verification!
   * ```
   */
  async create(params: {
    url: string;
    events: WebhookEventType[];
  }): Promise<WebhookEndpoint> {
    const response = await this.http.post<Record<string, unknown>>('/webhooks', params);
    return this.transformEndpoint(response);
  }

  /**
   * List all webhook endpoints.
   *
   * @returns List of webhook endpoints
   */
  async list(): Promise<ListResponse<WebhookEndpoint>> {
    const response = await this.http.get<{
      data?: Record<string, unknown>[];
      items?: Record<string, unknown>[];
      webhooks?: Record<string, unknown>[];
      total: number;
      hasMore?: boolean;
      has_more?: boolean;
    }>('/webhooks');

    const items = response.data || response.items || response.webhooks || [];

    return {
      data: items.map(e => this.transformEndpoint(e)),
      total: response.total || items.length,
      hasMore: response.hasMore || response.has_more || false,
    };
  }

  /**
   * Get a webhook endpoint by ID.
   *
   * @param webhookId - Webhook ID
   * @returns Webhook endpoint
   */
  async get(webhookId: string): Promise<WebhookEndpoint> {
    const response = await this.http.get<Record<string, unknown>>(
      `/webhooks/${webhookId}`
    );
    return this.transformEndpoint(response);
  }

  /**
   * Update a webhook endpoint.
   *
   * @param webhookId - Webhook ID
   * @param params - Update parameters
   * @returns Updated webhook endpoint
   */
  async update(
    webhookId: string,
    params: { url?: string; events?: WebhookEventType[]; active?: boolean }
  ): Promise<WebhookEndpoint> {
    const response = await this.http.put<Record<string, unknown>>(
      `/webhooks/${webhookId}`,
      params
    );
    return this.transformEndpoint(response);
  }

  /**
   * Delete a webhook endpoint.
   *
   * @param webhookId - Webhook ID
   */
  async delete(webhookId: string): Promise<void> {
    await this.http.delete(`/webhooks/${webhookId}`);
  }

  /**
   * Rotate webhook secret.
   *
   * @param webhookId - Webhook ID
   * @returns New secret
   */
  async rotateSecret(webhookId: string): Promise<{ secret: string }> {
    return this.http.post(`/webhooks/${webhookId}/rotate-secret`);
  }

  /**
   * Verify webhook signature using HMAC-SHA256.
   * Use this in your webhook handler to verify the request came from Execution Market.
   *
   * @param payload - Raw request body as string
   * @param signature - Value of X-EM-Signature header
   * @param secret - Webhook secret
   * @returns True if signature is valid
   *
   * @example
   * ```typescript
   * // Express.js example
   * app.post('/webhooks/execution-market', express.raw({ type: 'application/json' }), (req, res) => {
   *   const signature = req.headers['x-em-signature'] as string;
   *   const payload = req.body.toString();
   *
   *   if (!webhooks.verifySignature(payload, signature, WEBHOOK_SECRET)) {
   *     return res.status(401).send('Invalid signature');
   *   }
   *
   *   const event = JSON.parse(payload);
   *   // Process event...
   *   res.status(200).send('OK');
   * });
   * ```
   */
  verifySignature(payload: string, signature: string, secret: string): boolean {
    // Dynamic import to support both Node.js and browser environments
    try {
      // Node.js environment
      const crypto = require('crypto');
      const expected = crypto
        .createHmac('sha256', secret)
        .update(payload)
        .digest('hex');

      const expectedSignature = `sha256=${expected}`;

      // Use timing-safe comparison to prevent timing attacks
      if (signature.length !== expectedSignature.length) {
        return false;
      }

      return crypto.timingSafeEqual(
        Buffer.from(signature),
        Buffer.from(expectedSignature)
      );
    } catch {
      // Browser environment - use SubtleCrypto
      // Note: This is synchronous fallback; for async use verifySignatureAsync
      console.warn(
        'verifySignature requires Node.js crypto module. ' +
        'Use verifySignatureAsync for browser environments.'
      );
      return false;
    }
  }

  /**
   * Verify webhook signature asynchronously (works in browsers).
   *
   * @param payload - Raw request body as string
   * @param signature - Value of X-EM-Signature header
   * @param secret - Webhook secret
   * @returns Promise resolving to true if signature is valid
   */
  async verifySignatureAsync(
    payload: string,
    signature: string,
    secret: string
  ): Promise<boolean> {
    const encoder = new TextEncoder();
    const key = await crypto.subtle.importKey(
      'raw',
      encoder.encode(secret),
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign']
    );

    const signatureBytes = await crypto.subtle.sign(
      'HMAC',
      key,
      encoder.encode(payload)
    );

    const expected = `sha256=${Array.from(new Uint8Array(signatureBytes))
      .map(b => b.toString(16).padStart(2, '0'))
      .join('')}`;

    // Timing-safe comparison
    if (signature.length !== expected.length) {
      return false;
    }

    let result = 0;
    for (let i = 0; i < signature.length; i++) {
      result |= signature.charCodeAt(i) ^ expected.charCodeAt(i);
    }
    return result === 0;
  }

  /**
   * Transform API response to WebhookEndpoint object.
   */
  private transformEndpoint(raw: Record<string, unknown>): WebhookEndpoint {
    return {
      id: raw.id as string,
      url: raw.url as string,
      events: raw.events as WebhookEventType[],
      active: (raw.active ?? true) as boolean,
      secret: raw.secret as string,
      createdAt: new Date((raw.created_at ?? raw.createdAt) as string),
    };
  }
}

File: em-plugin-sdk/em_plugin_sdk/resources/webhooks.py
"""Webhooks resource — client.webhooks.create(), .list(), .verify_signature()."""

from __future__ import annotations

import hashlib
import hmac
from typing import Any, TYPE_CHECKING

from ..models import Webhook, WebhookList

if TYPE_CHECKING:
    from ..client import EMClient


class WebhooksResource:
    """Webhook CRUD and signature verification.

    Usage::

        wh = await client.webhooks.create(url="https://my.app/em-hook", events=["task.completed"])
        hooks = await client.webhooks.list()
        await client.webhooks.delete(wh.id)

        # Verify incoming webhook
        is_valid = client.webhooks.verify_signature(body, signature, timestamp, secret)
    """

    def __init__(self, client: EMClient) -> None:
        self._client = client

    async def create(
        self,
        url: str,
        events: list[str],
        *,
        description: str | None = None,
    ) -> Webhook:
        """Register a new webhook endpoint."""
        body: dict[str, Any] = {"url": url, "events": events}
        if description:
            body["description"] = description
        data = await self._client._request("POST", "/webhooks/", json=body)
        return Webhook.model_validate(data)

    async def list(self) -> WebhookList:
        """List all webhooks for the authenticated agent."""
        data = await self._client._request("GET", "/webhooks/")
        return WebhookList.model_validate(data)

    async def get(self, webhook_id: str) -> Webhook:
        """Get a single webhook by ID."""
        data = await self._client._request("GET", f"/webhooks/{webhook_id}")
        return Webhook.model_validate(data)

    async def update(
        self,
        webhook_id: str,
        *,
        url: str | None = None,
        events: list[str] | None = None,
        active: bool | None = None,
    ) -> Webhook:
        """Update a webhook."""
        body: dict[str, Any] = {}
        if url is not None:
            body["url"] = url
        if events is not None:
            body["events"] = events
        if active is not None:
            body["active"] = active
        data = await self._client._request("PUT", f"/webhooks/{webhook_id}", json=body)
        return Webhook.model_validate(data)

    async def delete(self, webhook_id: str) -> dict[str, Any]:
        """Delete a webhook."""
        return await self._client._request("DELETE", f"/webhooks/{webhook_id}")

    async def rotate_secret(self, webhook_id: str) -> dict[str, Any]:
        """Rotate the signing secret for a webhook."""
        return await self._client._request("POST", f"/webhooks/{webhook_id}/rotate-secret")

    async def test(self, webhook_id: str) -> dict[str, Any]:
        """Send a test ping to a webhook endpoint."""
        return await self._client._request("POST", f"/webhooks/{webhook_id}/test")

    @staticmethod
    def verify_signature(
        body: bytes,
        signature: str,
        timestamp: str,
        secret: str,
    ) -> bool:
        """Verify an incoming webhook's HMAC-SHA256 signature.

        Args:
            body: Raw request body bytes.
            signature: Value of X-EM-Signature header.
            timestamp: Value of X-EM-Timestamp header.
            secret: Webhook signing secret.

        Returns:
            True if the signature is valid.
        """
        expected = hmac.new(
            secret.encode(),
            f"{timestamp}.".encode() + body,
            hashlib.sha256,
        ).hexdigest()
        return hmac.compare_digest(expected, signature)
